File: examples/OBB.py
import numpy as np
from quanestimation import *
from scipy.integrate import simps

#initial state
rho0 = 0.5*np.array([[1., 1.],[1., 1.]])
#Hamiltonian
B = np.pi/2.0
sx = np.array([[0., 1.],[1., 0.0j]])
sy = np.array([[0., -1.j],[1.j, 0.]]) 
sz = np.array([[1., 0.0j],[0., -1.]])
H0_func = lambda x: 0.5*B*(sx*np.cos(x)+sz*np.sin(x))
dH_func = lambda x: [0.5*B*(-sx*np.sin(x)+sz*np.cos(x))]
d2H_func = lambda x: [0.5*B*(-sx*np.cos(x)-sz*np.sin(x))]
tspan = np.linspace(0., 1.0, 1000)
#prior distribution
x = [np.linspace(-0.5*np.pi, 0.5*np.pi, 100)]
mu, eta = 0.0, 0.5
eta_span = np.arange(0.2, 5.1, 0.1)
p_func = lambda x: np.exp(-(x-mu)**2/(2*eta**2))/(eta*np.sqrt(2*np.pi))
dp_func = lambda x: -(x-mu)*np.exp(-(x-mu)**2/(2*eta**2))/(eta**3*np.sqrt(2*np.pi))
p_tp = [p_func(x[0][i]) for i in range(len(x[0]))]
dp_tp = [dp_func(x[0][i]) for i in range(len(x[0]))]
c = simps(p_tp, x[0])
p = p_tp/c
dp = dp_tp/c
rho = [[] for i in range(len(x[0]))]
drho = [[] for i in range(len(x[0]))]
d2rho = [[] for i in range(len(x[0]))]
for i in range(len(x[0])):
    H0_tp = H0_func(x[0][i])
    dH_tp = dH_func(x[0][i])
    d2H_tp = d2H_func(x[0][i])
    dynamics = Lindblad(tspan, rho0, H0_tp, dH_tp)
    rho_tp, drho_tp, d2rho_tp = dynamics.secondorder_derivative(d2H_tp)
    rho[i] = rho_tp
    drho[i] = drho_tp
    d2rho[i] = d2rho_tp
f_OBB = OBB(x, p, dp, rho, drho, d2rho)
print(f_OBB)

File: quanestimation/examples/Julia examples/OBB.jl
using Trapz
include("../src/QuanEstimation.jl")

# initial state
rho0 = 0.5*[1.0 1.0+0.0im; 1.0 1.0]
# free Hamiltonian
B = pi/2.0
sx = [0.0 1.0; 1.0 0.0im]
sy = [0.0 -1.0im; 1.0im 0.0]
sz = [1.0 0.0im; 0.0 -1.0]
function H0_func(x)
    return 0.5*B*(sx*cos(x)+sz*sin(x))
end
function dH_func(x)
    return [0.5*B*(-sx*sin(x)+sz*cos(x))]
end
function d2H_func(x)
    return [0.5*B*(-sx*cos(x)-sz*sin(x))]
end
# dynamics
decay_opt = [zeros(ComplexF64,size(rho0)[1],size(rho0)[1])] 
gamma = [0.0]
tspan = range(0.0, stop=1.0, length=1000)
# prior distribution
function p_func(x, mu, eta)
    return exp(-(x-mu)^2/(2*eta^2))/(eta*sqrt(2*pi))
end
function dp_func(x, mu, eta)
    return -(x-mu)*exp(-(x-mu)^2/(2*eta^2))/(eta^3*sqrt(2*pi))
end
x = [range(-pi/2.0, stop=pi/2.0, length=100)].|>Vector
mu, eta = 0.0, 0.5
p_tp = [p_func(x[1][i], mu, eta) for i in 1:length(x[1])]
dp_tp = [dp_func(x[1][i], mu, eta) for i in 1:length(x[1])]
c = trapz(x[1], p_tp)
p, dp = p_tp/c, dp_tp/c

rho = Vector{Matrix{ComplexF64}}(undef, length(x[1]))
drho = Vector{Vector{Matrix{ComplexF64}}}(undef, length(x[1]))
d2rho = Vector{Vector{Matrix{ComplexF64}}}(undef, length(x[1]))
for i = 1:length(x[1]) 
    H0_tp = H0_func(x[1][i])
    dH_tp = dH_func(x[1][i])
    d2H_tp = d2H_func(x[1][i])
    rho_tp, drho_tp, d2rho_tp = QuanEstimation.secondorder_derivative(H0_tp, dH_tp, d2H_tp, rho0, decay_opt, gamma, 
                      [zeros(ComplexF64,size(rho0)[1],size(rho0)[1])], [zeros(length(tspan)-1)],tspan)
    rho[i] = rho_tp
    drho[i] = drho_tp
    d2rho[i] = d2rho_tp
end
f_OBB = QuanEstimation.OBB(x, p, dp, rho, drho, d2rho)
println(f_OBB)